server/src/compat/index_rebuilding/index_rebuilder.rs

```diff
@@ -0,0 +1,91 @@
+use std::io::SeekFrom;
+
+use crate::streaming::segments::storage::INDEX_SIZE;
+use crate::streaming::utils::file;
+use crate::{
+    server_error::CompatError, streaming::batching::message_batch::RETAINED_BATCH_OVERHEAD,
+};
+use bytes::{BufMut, BytesMut};
+use tokio::io::{AsyncReadExt, AsyncSeekExt, AsyncWriteExt, BufReader, BufWriter};
+
+const BUFFER_SIZE: usize = 1024 * 1024; // 1MB buffer
+const INDEX_BATCH_SIZE: usize = 1000 * INDEX_SIZE as usize;
+
+// Same struct as RetainedMessageBatch, but without payload
+pub struct BatchHeader {
+    base_offset: u64,
+    last_offset_delta: u32,
+    max_timestamp: u64,
+    length: u32,
+}
+
+pub struct IndexRebuilder {
+    pub log_path: String,
+    pub index_path: String,
+}
+
+impl IndexRebuilder {
+    pub fn new(log_path: String, index_path: String) -> Self {
+        Self {
+            log_path,
+            index_path,
+        }
+    }
+
+    async fn read_batch_header(
+        reader: &mut BufReader<tokio::fs::File>,
+    ) -> Result<Option<BatchHeader>, CompatError> {
+        Ok(match reader.read_u64_le().await {
+            Ok(base_offset) => Some(BatchHeader {
+                base_offset,
+                length: reader.read_u32_le().await?,
+                last_offset_delta: reader.read_u32_le().await?,
+                max_timestamp: reader.read_u64_le().await?,
+            }),
+            Err(_) => None,
```

Comment:
Why return Option from this function ? is it to handle EOF ?, maybe move it outside of this function